File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/digitaltwin/digitaltwin/DigitalTwinRepositoryType.java
package org.opcfoundation.webserver.digitaltwin.digitaltwin;

import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.opcfoundation.webserver.addressspace.reactiveobject.UaReactiveObjectType;
import org.opcfoundation.webserver.addressspace.nodes.UaNode;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaObjectTypes;
import org.opcfoundation.webserver.digitaltwin.DigitalTwinSpace;
import org.opcfoundation.webserver.digitaltwin.callback.DigitalTwinDirectoryCallback;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListResponse;
import org.opcfoundation.webserver.service.message.reactiveobject.BrowseObjectRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.BrowseObjectResponse;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadObjectAttributeRequest;
import org.opcfoundation.webserver.service.message.reactiveobject.ReadObjectAttributeResponse;
import org.opcfoundation.webserver.types.common.UaBrowseAdditionalInfo;
import org.opcfoundation.webserver.types.common.UaInstanceIdentifier;
import org.opcfoundation.webserver.types.common.UaObjectIdentifier;
import org.opcfoundation.webserver.types.common.UaReferenceDescriptor;
import org.opcfoundation.webserver.types.digitaltwin.DigitalTwinDescriptor;
import org.opcfoundation.webserver.types.digitaltwin.ObjectServiceContext;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public abstract class DigitalTwinRepositoryType extends UaReactiveObjectType implements DigitalTwinDirectoryCallback {
    public DigitalTwinRepositoryType(String typeId,
                                     LocalizedText displayName,
                                     DigitalTwinSpace twinSpace)
    {
        super(typeId, displayName, UaObjectTypes.DigitalTwinRepositoryType, twinSpace);
    }

    @Override
    public final CompletableFuture<BrowseObjectResponse> onBrowseObjectChildren(BrowseObjectRequest request)
    {
        if (request.getBrowseDescription().getReferenceTypeId().equals(NodeIds.HasComponent) ||
                request.getBrowseDescription().getReferenceTypeId().equals(NodeIds.Aggregates) ||
                !request.getAdditionalInfo().isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK))
            return CompletableFuture.completedFuture(new BrowseObjectResponse(new ArrayList<>(), false));

        ObjectServiceContext context = new ObjectServiceContext(request.getObjectId());
        GetDigitalTwinListRequest getElementListRequest= new GetDigitalTwinListRequest(
                context,
                request.getAdditionalInfo().getMaxReferencesPerNode(),
                request.getAdditionalInfo().getReferenceOffset());

        return onGetDigitalTwinList(getElementListRequest).
                thenApply(this::processBrowseObjectChildrenResponse);
    }

    @Override
    public final CompletableFuture<ReadObjectAttributeResponse> onReadObjectAttributes(ReadObjectAttributeRequest request) {
        UaInstanceIdentifier objectIdentifier = new UaInstanceIdentifier(
                new UaObjectIdentifier(nodeId().toParseableString(), request.getObjectId().getId(), null),
                null);

        NodeId directoryId = new NodeId(nodeManager.nsIndex(), objectIdentifier.toByteString());
        UaNode directoryNode = nodeManager.getNode(directoryId);

        if (null == directoryNode || NodeClass.Object != directoryNode.nodeClass()) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

        return CompletableFuture.completedFuture(new ReadObjectAttributeResponse(
                request.getObjectId().getId(),
                directoryNode.displayName(),
                directoryNode.description()));
    }

    private BrowseObjectResponse processBrowseObjectChildrenResponse(GetDigitalTwinListResponse response)
    {
        List<UaReferenceDescriptor> childDescriptors = new ArrayList<>();

        for (DigitalTwinDescriptor item: response.getDigitalTwins())
        {
            UaReferenceDescriptor descriptor = new UaReferenceDescriptor(
                    item.getId(),
                    NodeClass.Object,
                    item.getId(),
                    item.getDisplayName(),
                    item.getTypeId(),
                    NodeIds.Organizes,
                    false);

            childDescriptors.add(descriptor);
        }

        return new BrowseObjectResponse(childDescriptors, response.containsMoreData());
    }
}

File: opcua-webservice-node/src/server/digitaltwin/digitaltwin/DigitalTwinRepositoryType.ts
import { NodeClass, StatusCodes } from "opcua-webapi";
import { makeUaStatusCode, ReferenceTypeIds, UaError, UaLocalizedText, UaNodeId } from "opcua-webapi-ts";
import { UaReactiveObjectType } from "../../addressspace/reactiveobject/UaReactiveObjectType";
import { UaObjectTypes } from "../../addressspace/nodes/builtin";
import { DigitalTwinDirectoryCallback } from "../callback/DigitalTwinDirectoryCallback";
import {
    BrowseObjectRequest,
    BrowseObjectResponse,
    GetDigitalTwinListRequest,
    GetDigitalTwinListResponse,
    ReadObjectAttributeRequest,
    ReadObjectAttributeResponse,
} from "../../service/message";
import { UaBrowseAdditionalInfo, UaInstanceIdentifier, UaObjectIdentifier, UaReferenceDescriptor } from "../../types";
import { ObjectServiceContext } from "../../types/digitaltwin/ObjectServiceContext";
import { DigitalTwinSpace } from "../DigitalTwinSpace";

export abstract class DigitalTwinRepositoryType extends UaReactiveObjectType implements DigitalTwinDirectoryCallback {
    constructor(
        typeId: string,
        displayName: UaLocalizedText,
        twinSpace: DigitalTwinSpace) {
        super(typeId, displayName, UaObjectTypes.DigitalTwinRepositoryType, twinSpace);
    }

    digitalTwinSpace(): DigitalTwinSpace {
        return this.nodeManager as DigitalTwinSpace;
    }

    // Override to provide digital twin list
    abstract onGetDigitalTwinList(request: GetDigitalTwinListRequest): Promise<GetDigitalTwinListResponse>;

    // Implementation of parent type methods, don't override 
    override async onBrowseObjectChildren(request: BrowseObjectRequest): Promise<BrowseObjectResponse> {
        const referenceTypeId = request.browseDescription.referenceTypeId;

        if (!request.additionalInfo.isTaskRequired(UaBrowseAdditionalInfo.GET_CHILD_OBJECT_TASK) ||            
            (!referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.HierarchicalReferences)) &&
             !referenceTypeId.equal(UaNodeId.from(ReferenceTypeIds.Organizes))))
        {
            return new BrowseObjectResponse([], false);
        }

        const context = new ObjectServiceContext(request.objectId);
        const response = await this.onGetDigitalTwinList(
            new GetDigitalTwinListRequest(
                context,
                request.additionalInfo.maxReferencesPerNode,
                request.additionalInfo.referenceOffset,
            ),
        );

        return this.processBrowseObjectChildrenResponse(response);
    }

    // Implementation of parent type methods, don't override 
    override async onReadObjectAttributes(request: ReadObjectAttributeRequest): Promise<ReadObjectAttributeResponse> {
        const objectIdentifier = new UaInstanceIdentifier(
            new UaObjectIdentifier(this.nodeId.toString(), request.objectId.id, null),
            null,
        );
        const directoryId = new UaNodeId(objectIdentifier.toByteString(), this.nodeManager.nsIndex());
        const directoryNode = this.nodeManager.getNode(directoryId);

        if (directoryNode === null || directoryNode.nodeClass !== NodeClass.Object) {
            throw new UaError(makeUaStatusCode(StatusCodes.BadNodeIdUnknown));
        }

        return new ReadObjectAttributeResponse(
            request.objectId.id,
            directoryNode.displayName,
            directoryNode.description,
        );
    }

    private processBrowseObjectChildrenResponse(response: GetDigitalTwinListResponse): BrowseObjectResponse {
        const childDescriptors: Array<UaReferenceDescriptor> = [];

        for (const item of response.digitalTwins) {
            childDescriptors.push(
                new UaReferenceDescriptor(
                    item.id,
                    NodeClass.Object,
                    item.id,
                    item.displayName,
                    item.typeId,
                    UaNodeId.from(ReferenceTypeIds.Organizes),
                    false,
                ),
            );
        }

        return new BrowseObjectResponse(childDescriptors, response.containsMoreData);
    }
}